File: cli-node/src/commands/annotate.ts
import { resolve } from 'node:path';
import { GitAnalyzer } from '../core/git-analyzer.js';
import { buildResult } from '../core/timeline.js';
import { insertAnnotation } from '../annotator/inserter.js';
import { AnnotationOptions } from '../types.js';

export interface AnnotateOptions {
  style: 'brief' | 'detailed';
}

export async function annotate(file: string, lineRange: string, options: AnnotateOptions): Promise<void> {
  const [startStr, endStr] = lineRange.split('-');
  const startLine = parseInt(startStr);
  const endLine = endStr ? parseInt(endStr) : startLine;

  if (isNaN(startLine) || isNaN(endLine)) {
    console.error('Invalid line range. Use format: 10-50 or single line: 10');
    process.exit(1);
  }

  const filePath = resolve(file);
  const analyzer = new GitAnalyzer(process.cwd());

  if (!(await analyzer.isGitRepo())) {
    console.error('Not a git repository. Run this command inside a git project.');
    process.exit(1);
  }

  const commits = await analyzer.getLineHistory(file, startLine, endLine);

  if (commits.length === 0) {
    console.log('No git history found for this code segment.');
    return;
  }

  const result = buildResult(file, startLine, endLine, commits);
  const annotationOptions: AnnotationOptions = {
    style: options.style,
    includeAuthors: true,
    includeIssues: true,
  };

  const { success, message } = insertAnnotation(filePath, result, annotationOptions);
  if (success) {
    console.log(`✅ ${message}`);
  } else {
    console.error(`❌ ${message}`);
    process.exit(1);
  }
}

File: cli-python/src/code_archaeology/commands/annotate.py
import os

import click
from rich.console import Console

from ..core.git_analyzer import GitAnalyzer
from ..core.timeline import build_result
from ..annotator.inserter import insert_annotation

console = Console()


@click.command()
@click.argument("file")
@click.argument("line_range")
@click.option("--style", type=click.Choice(["brief", "detailed"]), default="detailed", help="Annotation style")
def annotate(file: str, line_range: str, style: str):
    """Insert a history annotation directly."""
    parts = line_range.split("-")
    start_line = int(parts[0])
    end_line = int(parts[1]) if len(parts) > 1 else start_line

    analyzer = GitAnalyzer(os.getcwd())

    if not analyzer.is_git_repo():
        console.print("[red]Not a git repository. Run this command inside a git project.[/]")
        raise SystemExit(1)

    commits = analyzer.get_line_history(file, start_line, end_line)

    if not commits:
        console.print("No git history found for this code segment.")
        return

    result = build_result(file, start_line, end_line, commits)
    file_path = os.path.abspath(file)

    success, message = insert_annotation(file_path, result, style)
    if success:
        console.print(f"[green]✅ {message}[/]")
    else:
        console.print(f"[red]❌ {message}[/]")
        raise SystemExit(1)
